README.md
Hybris
------

This is a custom version of the upstream project hosted at https://github.com/stskeeps/libhybris.git

Goal is to be merged with upstream asap.

Layout
------

- compat: For building on Android and also holds the common compat headers.
- hybris: The hybris bits to land on Ubuntu (using the common compat headers).


Building
--------

Type make


Notes
-----

Currently tested on the following hardware:

Galaxy Nexus (JB)
HTC One X (ICS)
Nexus 4 (JB)
Nexus 7 (JB)
Nexus 10 (JB)


debian/changelog
libhybris (0.60-0phablet1) quantal; urgency=low

  * Updating licenses and packaging to better fit Ubuntu's standards
    (LP: #1097083)

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Sun, 17 Feb 2013 23:58:15 -0300

libhybris (0.59) quantal; urgency=low

  * Removing extra debug from linker in case the library is not prelinked
    (disabled at recent Android versions)
  * Making libhybris buildable x86 64 (useful to avoid dependencies issues
    for packages depending but not using hybris at run time)

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Sun, 10 Feb 2013 23:36:26 -0200

libhybris (0.58) quantal; urgency=low

  * Removing code from ubuntu platform api, as it's now available at a
    different package (ubuntu-platform-api).

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Tue, 05 Feb 2013 04:53:46 -0200

libhybris (0.57) quantal; urgency=low

  * Place OSK input grab window according to coordinates and not only size

 -- Ricardo Mendoza <ricardo.mendoza@canonical.com>  Mon, 04 Feb 2013 13:33:25 -0430

libhybris (0.56) quantal; urgency=low

  * Correctly set window flags for input propagation

 -- Ricardo Mendoza <ricardo.mendoza@canonical.com>  Fri, 01 Feb 2013 03:39:45 -0430

libhybris (0.55) quantal; urgency=low

  [ Thomas Voss ]
  * Get rid of hard-coded sizes for input trap window.

  [ Ricardo Mendoza ]
  * Dynamically resize the OSK input window according to reported size 

 -- Ricardo Mendoza <ricardo.mendoza@canonical.com>  Thu, 31 Jan 2013 18:48:27 -0430

libhybris (0.54) quantal; urgency=low

  * Added mediaPrepared callback which signifies when libstagefright is ready
    to play the loaded media.

 -- Jim Hodapp <jim.hodapp@canonical.com>  Tue, 29 Jan 2013 15:46:08 -0500

libhybris (0.53) quantal; urgency=low

  [ Daniel d'Andrada ]
  * Added API to expose android's GPS through hybris

  [ Ricardo Salveti de Araujo ]
  * hybris/common: removing debugging prints by default

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Tue, 29 Jan 2013 08:59:18 -0500

libhybris (0.52) quantal; urgency=low

  * Added a function to initialize an ubuntu_sensor_observer instance for use by a client app.

 -- Jim Hodapp <jim.hodapp@canonical.com>  Thu, 24 Jan 2013 14:42:26 -0500

libhybris (0.51) quantal; urgency=low

  * Added blank / unblank methods to control power to display.

 -- Michael Frey <michael.frey@canonical.com>  Thu, 24 Jan 2013 09:36:45 -0500

libhybris (0.50) quantal; urgency=low

  * Added public API functions to enable/disable specific sensors.

 -- Jim Hodapp <jim.hodapp@canonical.com>  Wed, 23 Jan 2013 12:55:23 -0500

libhybris (0.49) quantal; urgency=low

  * Adding more stdio.h hooks to avoid calling __sfvwrite (malloc -> pthread)

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Wed, 23 Jan 2013 00:39:41 -0200

libhybris (0.48) quantal; urgency=low

  [ Daniel d'Andrada ]
  * Added hook for fseeko function.

  [ Ricardo Salveti de Araujo ]
  * Rebase compat code to work with Android's JB 4.2.1

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Tue, 22 Jan 2013 16:31:06 -0200

libhybris (0.47) quantal; urgency=low

  * Fix predicate when updating input configuration if osk/notifications visibility changes.

 -- Thomas Voß <thomas.voss@canonical.com>  Thu, 20 Dec 2012 17:15:24 +0100

libhybris (0.46) quantal; urgency=low

  * Add telephony to list of well known apps 

 -- Bill Filler <bill.filler@canonical.com>  Wed, 19 Dec 2012 19:21:58 -0500

libhybris (0.45) quantal; urgency=low

  * Fix input events being propagated through the osk to ordinary apps underneath.
 -- Thomas Voß <thomas.voss@canonical.com>  Tue, 18 Dec 2012 07:29:51 +0100

libhybris (0.44) quantal; urgency=low

  * Fix key event delivery and remove filtering of volume up key.

 -- Thomas Voß <thomas.voss@canonical.com>  Mon, 17 Dec 2012 15:31:36 +0100

libhybris (0.43) quantal; urgency=low

  * Bumped package version.

 -- Loïc Molinari <loic.molinari@canonical.com>  Mon, 17 Dec 2012 23:30:41 +0100

libhybris (0.42) quantal; urgency=low

  * Add a callback for session observers that signals when a session is unfocused.
  * Introduce surface flags and a way to express that a surface should be opaque.

 -- Thomas Voß <thomas.voss@canonical.com>  Fri, 14 Dec 2012 12:31:36 +0100

libhybris (0.41) quantal; urgency=low

  * Stop and continue ordinary app sessions.

 -- Thomas Voß <thomas.voss@canonical.com>  Fri, 14 Dec 2012 12:00:42 +0100

libhybris (0.40) quantal; urgency=low

  [ Thomas Voß ]
  * Do not focus system sessions.

 -- Michael Frey <michael.frey@canonical.com>  Thu, 13 Dec 2012 17:05:42 -0500

libhybris (0.39) quantal; urgency=low

  * Add a way to unfocus all running applications.

 -- Thomas Voß <thomas.voss@canonical.com>  Wed, 12 Dec 2012 14:49:36 +0100

libhybris (0.38) quantal; urgency=low

  * Fix event delivery for system-session only scenarios.
  * Add a new callback "on_session_requested" for signalling to observers: Please start a session with the following properties.

 -- Thomas Voß <thomas.voss@canonical.com>  Tue, 11 Dec 2012 19:31:51 +0100

libhybris (0.37) quantal; urgency=low

  * Iterate surface roles and adjust stacking.

 -- Thomas Voß <thomas.voss@canonical.com>  Thu, 06 Dec 2012 21:00:04 +0100

libhybris (0.36) quantal; urgency=low

  * Refactor sensors API to include proximity and ambient light readings.

 -- Thomas Voß <thomas.voss@canonical.com>  Wed, 05 Dec 2012 18:36:36 +0100

libhybris (0.35) quantal; urgency=low

  * Bumping the package version.

 -- Jim Hodapp <jim.hodapp@canonical.com>  Wed, 05 Dec 2012 11:54:39 -0500

libhybris (0.34) quantal; urgency=low

  * Do not restack surfaces on request update
  * Add query of screen dpi.

 -- Thomas Voß <thomas.voss@canonical.com>  Tue, 04 Dec 2012 20:53:26 +0100

libhybris (0.33) quantal; urgency=low

  [ Thomas Voß ]
  * Add a generic way for application sessions signalling
    an update to the appmanager.


hybris/ics/linker.h
/*
 * Copyright (C) 2008 The Android Open Source Project
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef _LINKER_H_
#define _LINKER_H_

#include <unistd.h>
#include <sys/types.h>
#include <elf.h>

#undef PAGE_MASK
#undef PAGE_SIZE
#define PAGE_SIZE 4096
#define PAGE_MASK 4095

#if defined(__x86_64__)
typedef Elf64_Ehdr Elf_Ehdr;
typedef Elf64_Shdr Elf_Shdr;
typedef Elf64_Sym  Elf_Sym;
typedef Elf64_Addr Elf_Addr;
typedef Elf64_Phdr Elf_Phdr;
typedef Elf64_Half Elf_Half;
typedef Elf64_Rel  Elf_Rel;
typedef Elf64_Rela Elf_Rela;
#else
typedef Elf32_Ehdr Elf_Ehdr;
typedef Elf32_Shdr Elf_Shdr;
typedef Elf32_Sym  Elf_Sym;
typedef Elf32_Addr Elf_Addr;
typedef Elf32_Phdr Elf_Phdr;
typedef Elf32_Half Elf_Half;
typedef Elf32_Rel  Elf_Rel;
typedef Elf32_Rela Elf_Rela;
#endif

void debugger_init();
const char *addr_to_name(unsigned addr);

/* magic shared structures that GDB knows about */

struct link_map
{
    uintptr_t l_addr;
    char * l_name;
    uintptr_t l_ld;
    struct link_map * l_next;
    struct link_map * l_prev;
};

/* needed for dl_iterate_phdr to be passed to the callbacks provided */
struct dl_phdr_info
{
    Elf_Addr dlpi_addr;
    const char *dlpi_name;
    const Elf_Phdr *dlpi_phdr;
    Elf_Half dlpi_phnum;
};


// Values for r_debug->state
enum {
    RT_CONSISTENT,
    RT_ADD,
    RT_DELETE
};

struct r_debug
{
    int32_t r_version;
    struct link_map * r_map;
    void (*r_brk)(void);
    int32_t r_state;
    uintptr_t r_ldbase;
};

typedef struct soinfo soinfo;

#define FLAG_LINKED     0x00000001
#define FLAG_ERROR      0x00000002
#define FLAG_EXE        0x00000004 // The main executable

#define SOINFO_NAME_LEN 128

struct soinfo
{
    const char name[SOINFO_NAME_LEN];
    Elf_Phdr *phdr;
    int phnum;
    unsigned entry;
    unsigned base;
    unsigned size;

    int unused;  // DO NOT USE, maintained for compatibility.

    unsigned *dynamic;

    unsigned wrprotect_start;
    unsigned wrprotect_end;

    soinfo *next;
    unsigned flags;

    const char *strtab;
    Elf_Sym *symtab;

    unsigned nbucket;
    unsigned nchain;
    unsigned *bucket;
    unsigned *chain;

    unsigned *plt_got;

    Elf_Rel *plt_rel;
    unsigned plt_rel_count;

    Elf_Rel *rel;
    unsigned rel_count;

#ifdef ANDROID_SH_LINKER
    Elf_Rela *plt_rela;
    unsigned plt_rela_count;

    Elf_Rela *rela;
    unsigned rela_count;
#endif /* ANDROID_SH_LINKER */

    unsigned *preinit_array;
    unsigned preinit_array_count;

    unsigned *init_array;
    unsigned init_array_count;
    unsigned *fini_array;
    unsigned fini_array_count;

    void (*init_func)(void);
    void (*fini_func)(void);

#ifdef ANDROID_ARM_LINKER
    /* ARM EABI section used for stack unwinding. */
    unsigned *ARM_exidx;
    unsigned ARM_exidx_count;
#endif

    unsigned refcount;
    struct link_map linkmap;
};


extern soinfo libdl_info;

#ifdef ANDROID_ARM_LINKER

#define R_ARM_COPY       20
#define R_ARM_GLOB_DAT   21
#define R_ARM_JUMP_SLOT  22
#define R_ARM_RELATIVE   23

/* According to the AAPCS specification, we only
 * need the above relocations. However, in practice,
 * the following ones turn up from time to time.
 */
#define R_ARM_ABS32      2
#define R_ARM_REL32      3

#elif defined(ANDROID_X86_LINKER)

#define R_386_32         1
#define R_386_PC32       2
#define R_386_GLOB_DAT   6
#define R_386_JUMP_SLOT  7
#define R_386_RELATIVE   8

#elif defined(ANDROID_SH_LINKER)

#define R_SH_DIR32      1
#define R_SH_GLOB_DAT   163
#define R_SH_JUMP_SLOT  164
#define R_SH_RELATIVE   165

#endif /* ANDROID_*_LINKER */


#ifndef DT_INIT_ARRAY
#define DT_INIT_ARRAY      25
#endif

#ifndef DT_FINI_ARRAY
#define DT_FINI_ARRAY      26
#endif

#ifndef DT_INIT_ARRAYSZ
#define DT_INIT_ARRAYSZ    27
#endif

#ifndef DT_FINI_ARRAYSZ
#define DT_FINI_ARRAYSZ    28
#endif

#ifndef DT_PREINIT_ARRAY
#define DT_PREINIT_ARRAY   32
#endif

#ifndef DT_PREINIT_ARRAYSZ
#define DT_PREINIT_ARRAYSZ 33
#endif

soinfo *find_library(const char *name);
unsigned unload_library(soinfo *si);
Elf_Sym *lookup_in_library(soinfo *si, const char *name);
Elf_Sym *lookup(const char *name, soinfo **found, soinfo *start);
soinfo *find_containing_library(const void *addr);
Elf_Sym *find_containing_symbol(const void *addr, soinfo *si);
const char *linker_get_error(void);

#ifdef ANDROID_ARM_LINKER 
typedef long unsigned int *_Unwind_Ptr;
_Unwind_Ptr dl_unwind_find_exidx(_Unwind_Ptr pc, int *pcount);
#elif defined(ANDROID_X86_LINKER) || defined(ANDROID_SH_LINKER)
int dl_iterate_phdr(int (*cb)(struct dl_phdr_info *, size_t, void *), void *);
#endif

#endif
